zeth-contracts/contracts/BaseMerkleTree.sol

```diff
@@ -7,76 +7,76 @@ pragma solidity ^0.5.0;
 // Adapted from: https://github.com/zcash-hackworks/babyzoe
 
 contract BaseMerkleTree {
-  // Index of the current node: Index to insert the next incoming commitment
-  uint currentNodeIndex;
+    // Depth of the merkle tree (should be set with the same depth set in the
+    // cpp prover)
+    uint constant depth = 4;
 
-  // Array containing the 2^(depth) leaves of the merkle tree
-  // We can switch the leaves to be of type bytes and not bytes32
-  // to support digest of various size (eg: if we use different hash functions)
-  // That way we'd have a merkle tree for any type of hash function (that can be implemented
-  // as a precompiled contract for instance)
-  //
-  // Leaves is a 2D array
-  bytes32[] leaves; // Declared as a dynamic array, the bound is put in the constructor
+    // Number of leaves
+    uint constant nbLeaves = 2**depth;
 
-  // Depth of the merkle tree (should be set with the same depth set in the cpp prover)
-  uint depth;
+    // Index of the current node: Index to insert the next incoming commitment
+    uint currentNodeIndex;
 
-  // Number of leaves
-  uint nbLeaves;
+    // Array containing the 2^(depth) leaves of the merkle tree.  We can switch
+    // the leaves to be of type bytes and not bytes32 to support digest of
+    // various size (eg: if we use different hash functions).  That way we'd
+    // have a merkle tree for any type of hash function (that can be implemented
+    // as a precompiled contract for instance)
+    //
+    // Leaves is a 2D array
+    bytes32[nbLeaves] leaves;
 
-  // Debug only
-  event LogDebug(bytes32 message);
+    // Debug only
+    event LogDebug(bytes32 message);
 
-  // Constructor
-  constructor(uint treeDepth) public {
-    depth = treeDepth;
-    nbLeaves = 2**depth;
-
-    // Initialize the merkle tree with zeroes in values
-    bytes32 zeroes;
-    for (uint i = 0; i < nbLeaves; i++) {
-      leaves.push(zeroes);
+    // Constructor
+    constructor(uint treeDepth) public {
+        require (
```

Comment:
Ideally, the depth constant would come from a single configuration location (e.g. zeth.h), but since it is hard coded here I kept this as a way to check (at deployment time) that the configuration in the contract matches the client / prover server.

Reply:
> since it is hard coded here I kept this as a way to check (at deployment time) that the configuration in the contract matches the client / prover server.

To make sure that we worked with the same merkle tree depth as the one set on the client (which needed to be set to the same value as the one used by the prover_server - ie. the one in `zeth.h`) we previously just set the value of the storage variable `depth` to the value passed to the constructor, it was not a constant (https://github.com/clearmatics/zeth/blob/pyclient-cli/zeth-contracts/contracts/BaseMerkleTree.sol#L23). Since setting the `depth` to be a constant (which value equals the depth set on the client and prover) would avoid some variable assignment, and remove an argument in the constructor, we could definitely do it. However for now, switching this var to being a constant and adding this extra check in the constructor seems quite odd to me.
What is the status of this PR? Is it WIP?

Reply:
> Since setting the depth to be a constant (which value equals the depth set on the client and prover) would avoid some variable assignment, and remove an argument in the constructor, we could definitely do it (ie. switch depth to a constant). However for now, switching this var to being a constant and adding this extra check in the constructor seems quite odd to me.

The purpose of having the depth as a constant is so that the data structure sizes are known at compile time.  This makes the storage structures and the memory structures much simpler and cheaper to access, and saves a lot of gas in the average mix call (by my measurement about 20,000 just for depth of 4).

Avoiding the assignment / argument is not really a goal in and of itself - it is unlikely to save a significant amount of gas in comparison.

So I think it's worth having the depth as a constant, but what I'm concerned about is that it's currently very easy for the contract, client and prover_server configs to get out of sync, and if we want to change a setting, there are several places that must be changed, and it's not trivial to know all of them.  Ideally, we'd get some error at compile time if there is a difference.  In the worst case, we would not get an error until someone tries to execute a mix.  Here, I've kept the depth parameter to the constructor as it has almost no cost and gives an error at deployment time if client and contract have different settings.  Not as good as an error at compile time, or the previous case of it just being fully adaptive, but it seems like a fair trade-off, given the cost savings.

Similar to the number of joinsplit inputs / outputs, eventually it would be nice if the constants could come from a sinlge configuration file shared with the client and prover #143.

> What is the status of this PR? Is it WIP?

I removed the WIP from this PR.  It doesn't represent every possible optimization, but it represents what I think is a decent first-pass, based on measuring where the highest costs are.  It makes sense to re-measure some of these changes, since Istanbul was introduced part-way through, but each change introduced here was measured to make a meaningful difference to cost.

Reply:
> The purpose of having the depth as a constant is so that the data structure sizes are known at compile time. This makes the storage structures and the memory structures much simpler and cheaper to access, and saves a lot of gas in the average mix call

Yes I agree with that, we should switch to `const` when possible to save gas.
What I didn't get, was the `require` on the depth in the constructor. That's weird to give the user the possibility to set input values to a function that crashes for all but one value baked in the contract.

You said:
> Here, I've kept the depth parameter to the constructor as it has almost no cost and gives an error at deployment time if client and contract have different settings. Not as good as an error at compile time, or the previous case of it just being fully adaptive, but it seems like a fair trade-off given the cost savings.

So it looks like you've done this on purpose to get an error if the config off-chain and the config on the contract diverge. I'm not entirely seduced by the idea though. This looks like a hack/tweak to keep around while we dev, but we need to find a better way to handle this, maybe with #143 as you suggested. Ultimately, nothing will prevent the client and prover_server configs to get out of sync. You can deploy a mixer and decide to change the constants on the prover_server which would "run out of sync" with the config of the contract you are interacting with.

Reply:
> Ultimately, nothing will prevent the client and prover_server configs to get out of sync. You can deploy a mixer and decide to change the constants on the prover_server which would "run out of sync" with the config of the contract you are interacting with.

Sure.  I'll remove the param if you still think that's better, but just to explain, catching the case where a client changes their config after the contract has been deployed is not a goal of this.  The point is to be able to catch a mismatch in config *somewhere* (ideally during deployment) with more than just an opaque message like "invalid poof".  I thought of it more as a sanity check, given that different deployments may have different configs.  But I guess you could think of it as being for development.

> we need to find a better way to handle this, maybe with #143 as you suggested

Yes, it would be nice to have a single config somewhere.  I've been looking into this further and there seems to be no nice / easy solution.  So we should probably discuss the options for #143.  In the meantime it would be nice to continue with the rest of this PR, so just let me know if you still prefer to remove `depth` here.